lib/sequent/core/helpers/message_router.rb

```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+module Sequent
+  module Core
+    module Helpers
+      class MessageRouter
+        NO_MATCH = -> { Set.new }
+
+        attr_reader :routes
+
+        def initialize
+          @routes = Hash.new { |h, k| h[k] = NO_MATCH.call }
+        end
+
+        ##
+        # Registers the given handler for the given message classes.
+        #
+        def register_messages(*message_classes, handler)
+          message_classes.each do |message_class|
+            @routes[message_class] << handler
+          end
+        end
+
+        ##
+        # Returns a set of handlers that match the given message, or an empty set when none match.
+        #
+        def match_message(message)
+          @routes
+            .reduce(NO_MATCH.call) do |memo, (message_class, handlers)|
+              memo = memo.merge(handlers) if message.is_a?(message_class)
```

Comment:
This is a breaking change so we should make this optional somehow. There are several options: 

1. Put behind a configuration settings
```ruby
 memo = if Sequent.configuration.use_is_a_matcher
  memo.merge(handlers) if message.is_a?(message_class)
else
  memo.merge(handlers) if message == message_class
end
```

2. Keep the current behavior the default and implement a `matcher`  option that  you can pass  in the `on` block like we discussed in #316 

I think I prefer option 2, to implement this without an extra config item,  and for now use the `is_a` as an extra argument when registering the event using the `on` bock.

Maybe there are other options?

Reply:
Some shower thoughts on alternatives:

```ruby
module Matchers
  EqualsClass = Struct.new(:expected_class, keyword_init: true) do
    def matches_message?(message)
      message.class == expected_class
    end
  end

  IsA = Struct.new(:expected_class, keyword_init: true) do
    def matches_message?(message)
      message.is_a?(expected_class)
    end
  end
end
```

```ruby
module MessageMatchers
  module ClassMethods
    def is_a?(message_class)
      Matchers::IsA.new(expected_class: message_class)
    end
  end

  def self.included(host_class)
    host_class.extend(ClassMethods)
  end
end

class MyHandler
  include MessageHandler
  include MessageMatchers

  on SpecificEvent, is_a?(EventModule) do |message|
    # ...
  end
end
```

And `SpecificEvent` will automatically be wrapped in a `Matchers::EqualsClass` for backward compatibility.

Alternative DSL:

```
class MyHandler
  include MessageHandler

  on SpecificEvent, is_a: EventModule do |message|
    # ...
  end
end
```

Where we convert the value of `is_a` to a `Matchers::IsA` (could also be an array for multiple like: `on SpecificEvent, is_a: [SomeModule, OtherModule]`).

We could also provide a built-in `HasAttrs` matcher, used like this:

```ruby
class MyHandler
  include MessageHandler

  on SpecificEvent, attrs: {aggregate_type: 'MyType'} do |message|
    # ...
  end
end
```

Either way this opens up extensibility for Sequent users, allowing them to write custom 'matchers', like so:

```ruby
class MyHandler
  include MessageHandler

  class CustomMatch
    def matches_messages?(message)
      # ...
    end
  end

  class << self
    def custom_match
      CustomMatch.new
    end
  end

  on custom_match do |message|
    # ...
  end
end
```

Reply:
I've let this have some more thoughts and I would prefer the following solution:

```ruby
class MyHandler
  include MessageHandler

  on MyEvent do |event|
    # ...
  end

  on is_a(MyModule) do |event|
    # ...
  end

  on has_attrs(OtherEvent, attr: 'value', nested: {numeric: greater_than(0)}) do |event|
    # ...
  end
end
```

Advantages:

- The `on` method  signature would change from `on(*event_classes)` to `on(*message_matchers)` and we can automatically wrap a `Class` argument in a `class_equals` matcher for full backward compatibility.
- Sequent can provide built-in matchers for common cases.
- Sequent users can write their own matchers for proprietary cases.
- Matchers can be nested for more complex cases.
- No extra configuration needed.

Reply:
Nice one indeed!